crates/uv/src/commands/project/lock.rs

```diff
@@ -462,19 +465,20 @@ async fn do_lock(
             }
 
             // When we run the same resolution from the lockfile again, we could get a different result the
-            // second time due to the preferences causing us to skip a fork point (see
-            // "preferences-dependent-forking" packse scenario). To avoid this, we store the forks in the
+            // second time due to the preferences causing us to skip a fork point (see the
+            // `preferences-dependent-forking` packse scenario). To avoid this, we store the forks in the
             // lockfile. We read those after all the lockfile filters, to allow the forks to change when
             // the environment changed, e.g. the python bound check above can lead to different forking.
-            let resolver_markers = ResolverMarkers::universal(if upgrade.is_all() {
-                // We're discarding all preferences, so we're also discarding the existing forks.
-                vec![]
-            } else {
-                existing_lock
-                    .as_ref()
-                    .map(|existing_lock| existing_lock.lock().fork_markers().to_vec())
-                    .unwrap_or_default()
-            });
+            let resolver_markers = ResolverMarkers::universal(
+                reusable_lock
+                    .map(|lock| lock.fork_markers().to_vec())
+                    .unwrap_or_else(|| {
+                        environments
+                            .cloned()
+                            .map(SupportedEnvironments::into_markers)
```

Comment:
This is actually critical: iirc the resolver currently assumes these are non-overlapping, and if they aren't we can create a conflicting resolution

Reply:
Wait, what? Is that true? Where does that assumption come into play?

Reply:
I think you're right that they need to be disjoint. We can either (1) require that the user provides disjoint markers, or (2) `and` with the negation of the preceding markers when constructing this.

Reply:
Putting this here also for reference: The problem is: If we resolve foo==1 for `(sys_platform == "linux" or sys_platform == "darwin")` and foo==2 for `(sys_platform == "linux" or sys_platform == "nt")`, which one would we install on linux?